### PR TITLE
warnings

### PR DESCRIPTION
**Pull request by stoeffel**
screen shot 2018-01-05 at 21 55 42

### CODE REVIEW

**Comment by stoeffel** [comment]
one more warning that could be helpful would be `NoExpectation`.

File: src/VerifyExamples/ExposedApi.elm [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    value
+        |> firstSubmatch moduleExposing
+        |> Maybe.map (toExposedDefinition functions)
+        |> Maybe.withDefault None
+
+
+toExposedDefinition : List String -> String -> ExposedApi
+toExposedDefinition functions match =
+    case match of
+        ".." ->
```
Comment (stoeffel) [inline src/VerifyExamples/ExposedApi.elm]: we are missing aliases and types in the case where a moudle exposes everything. maybe we should just change the warning in this case that a module should be explicit about it's exposed definitions.

**Comment by stoeffel** [comment]
We'll need to implement `--report` for elm-verify-examples, with this change:-(

**Comment by stoeffel** [comment]
@eeue56 any thoughts on this?

**Comment by eeue56** [comment]
We could generate examples _for_ the user, for example, if you have this function:

```
add : Int -> Int -> Int 
add x y = x + y
```

Then we can use the fuzzy number generators to create some random examples, for example:

```
add doesn't have any examples! We generated some for you:

>>> add 0 0
       0

>>> add 1 10
       11

>>> add -1 -1
       -2
```

Was that the kind of thoughts you wanted? :D Otherwise, I think it sounds like a good and useful idea, especially if you can fail on warnings.

File: bin/runner.js [hunk 2]
```diff
@@ -19,75 +20,92 @@ var init = function(args){
...
+            ignoredWarnings: ignoredWarnings(model.ignoreWarnings, moduleName)
+          }
+        );
+    });
+  });
+
+  var warnings = [];
+  app.ports.warn.subscribe(function(args) {
+    var moduleName = args[0];
+    var warningsForModule = args[1];
```
Comment (eeue56) [inline bin/runner.js]: 🎨 why not send out `Json.Encode.object [ ("moduleName", Json.Encode.string moduleName), ("warnings", Json.Encode.list <| List.map Json.Encode.string warnings)]` instead?

File: bin/runner.js [hunk 3]
```diff
@@ -19,75 +20,92 @@ var init = function(args){
...
+    if (warningsForModule.length === 0) return;
+    warnings.push({
+      moduleName: moduleName,
+      warnings: warningsForModule
```
Comment (eeue56) [inline bin/runner.js]: if you did the above, then you wouldn't need to reconstruct this object :)

File: src/VerifyExamples/Ast/Intermediate.elm [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+toIntermediateAst : String -> Maybe IntermediateAst
+toIntermediateAst =
+    oneOf
+        [ firstSubmatch newLineRegex >> Maybe.map (always NewLine)
+        , firstSubmatch importRegex >> Maybe.map (Expression ImportPrefix)
+        , firstSubmatch typeRegex >> Maybe.map (Expression TypePrefix)
+        , firstSubmatch expectationRegex >> Maybe.map (Expression ArrowPrefix)
+        , firstSubmatch localFunctionRegex >> Maybe.map toFunctionExpression
+        , firstSubmatch assertionRegex >> Maybe.map ExpressionNotPrefixed
```
Comment (eeue56) [inline src/VerifyExamples/Ast/Intermediate.elm]: :art: these could be refactored to be like `Regex -> (String -> IntermediateAst) -> Maybe IntermediateAst`

File: src/VerifyExamples/Ast/Intermediate.elm [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        ]
+
+
+toFunctionExpression : String -> IntermediateAst
+toFunctionExpression str =
+    firstSubmatch functionNameRegex str
+        |> Maybe.withDefault "no function name given!"
```
Comment (eeue56) [inline src/VerifyExamples/Ast/Intermediate.elm]: Is this possible?

File: src/VerifyExamples/Ast/Intermediate.elm [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+
+toAst : List IntermediateAst -> List Ast
+toAst =
+    List.Extra.groupWhileTransitively group
+        >> List.filterMap convertGroup
+
+
+group : IntermediateAst -> IntermediateAst -> Bool
+group x y =
```
Comment (eeue56) [inline src/VerifyExamples/Ast/Intermediate.elm]: ❓ this name seems misleading -- what is it grouping?

File: src/VerifyExamples/Ast/Intermediate.elm [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+                TypePrefix ->
+                    notPrefixed y
+
+                ArrowPrefix ->
+                    arrowPrefixed y
+
+
+notPrefixed : IntermediateAst -> Bool
+notPrefixed ast =
```
Comment (eeue56) [inline src/VerifyExamples/Ast/Intermediate.elm]: Flip the name? `isPrefixed`?

File: src/VerifyExamples/Ast/Intermediate.elm [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    case ast of
+        ExpressionNotPrefixed _ ->
+            True
+
+        _ ->
+            False
+
+
+arrowPrefixed : IntermediateAst -> Bool
+arrowPrefixed ast =
```
Comment (eeue56) [inline src/VerifyExamples/Ast/Intermediate.elm]: :art: `isArrowPrefixed`

File: src/VerifyExamples/Ast/Intermediate.elm [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    case ast of
+        Expression ArrowPrefix _ ->
+            True
+
+        _ ->
+            False
+
+
+convertGroup : List IntermediateAst -> Maybe Ast
+convertGroup ast =
```
Comment (eeue56) [inline src/VerifyExamples/Ast/Intermediate.elm]: Convert a group of what to what?

File: src/VerifyExamples/ExposedApi.elm [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+        All ->
+            []
+
+        None ->
+            []
+
+
+everythingExposed : ExposedApi -> Bool
+everythingExposed api =
```
Comment (eeue56) [inline src/VerifyExamples/ExposedApi.elm]: 🎨 `isEverythingExposed`

File: src/VerifyExamples/ExposedApi.elm [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    case api of
+        Explicitly _ ->
+            False
```
Comment (eeue56) [inline src/VerifyExamples/ExposedApi.elm]: What if every name is exposed, just manually?

File: src/VerifyExamples/Warning.elm [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        |> ExposedApi.definitions
+        |> Maybe.Util.fromList
+
+
+notEverythingExposed : HasWarning
+notEverythingExposed { exposedApi } _ =
+    if ExposedApi.everythingExposed exposedApi then
+        Just []
+    else
+        Nothing
```
Comment (eeue56) [inline src/VerifyExamples/Warning.elm]: I'm finding things in this file are pretty hard to understand

File: src/VerifyExamples/ExposedApi.elm (hunk 11 above)
Comment (stoeffel) [reply src/VerifyExamples/ExposedApi.elm]: than it should still be this case

File: src/VerifyExamples/Warning.elm (hunk 12 above)
Comment (stoeffel) [reply src/VerifyExamples/Warning.elm]: yes, I know 😢. I just made it work, but didn't have time to refactor it yet. will let you know when you could review a cleaner version.

**Comment by eeue56** [comment]
I had one thought: what if someone doesn't want to give an example for a value? For example, a lot of the things in `elm-lang/html` or `elm-css` don't have examples. Examples would be great, I think, but what if they don't want them?

**Comment by stoeffel** [comment]
@eeue56 there is a way to ignore a warning completely for a module, not just for a function.
I wanna avoid adding something special in the comment to suppress warnings. Maybe it should just be a warning for definitions that have examples, but that are not verified?

File: bin/runner.js [hunk 13]
```diff
@@ -98,16 +111,42 @@ function cleanup(model) {
...
+    warn(chalk.underline("Warnings in module " + warnings.moduleName));
+    warn("\n");
+    warn(warnings.warnings.map(indent).join("\n"));
+    warn("\n");
+  };
+}
+
+function warnSummary(model, warnings) {
+  if (!model.warn) return;
+  var count = warnings.reduce(function(acc, warning) { return warning.warnings.length + acc; }, 0);
```
Comment (eeue56) [inline bin/runner.js]: maybe pull this out into a named function

**Comment by stoeffel** [comment]
I changed this to not show warnings by default (like elm-make) and since this is still in an early phase and I would like to add more warnings (or different in a future pr). also #58 is based on this and I would like to unblock that since it's a really nice feature.

File: src/VerifyExamples/Warning.elm [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+toString (Warning type_ definitions) =
+    case type_ of
+        NoExampleForExposedDefinition ->
+            "The following exposed definitions don't have examples:\n"
+                :: List.map ((++) "    - ") definitions
+                |> unlines
+
+        ExposingDotDot ->
+            "It's recommended to be specific about your exports.\n"
+                :: [ "    Don't use `module ... exposing (..)`"
```
Comment (eeue56) [inline src/VerifyExamples/Warning.elm]: Why use `::` here?

Comment (stoeffel) [reply src/VerifyExamples/Warning.elm]: oops.

File: example/tests/elm-verify-examples.json [hunk 15]
```diff
@@ -6,5 +6,20 @@
...
     "Failing",
     "Todo"
-  ]
+  ],
+  "ignoreWarnings": {
+    "Mock.Foo.Bar.Moo": [
+      {
+        "name": "dontCare",
+        "ignore": [
+          "NoExampleForExposedDefinition"
```
Comment (eeue56) [inline example/tests/elm-verify-examples.json]: So verbose to write this by hand. Wonder if we could do some magic like adding the ability to modify this from the cli. Like `elm-verify-examples --ignore Mock.Foo.Bar.Moo`

File: src/Json/Util.elm [hunk 16]
```diff
@@ -0,0 +1,17 @@
+module Json.Util exposing (exact)
+
+import Json.Decode as Decode exposing (Decoder)
+
+
+{-| A decoder that only succeeds when it decodes to the given value.
+-}
+exact : Decoder a -> a -> Decoder a
+exact decoder x =
```
Comment (eeue56) [inline src/Json/Util.elm]: `onlyIfExactly`?

Comment (eeue56) [reply src/Json/Util.elm]: `Json.Decode.onlyIfExactly`

File: src/Json/Util.elm [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+    Decode.andThen
+        (\y ->
+            if x == y then
+                Decode.succeed x
+            else
+                Decode.fail ("expected " ++ toString x ++ " got " ++ toString y)
```
Comment (eeue56) [inline src/Json/Util.elm]: comma after the `toString x`?

File: src/VerifyExamples/ExposedApi.elm (hunk 1 above)
Comment (eeue56) [reply src/VerifyExamples/ExposedApi.elm]: 👍

**Pull request merged**